src/ai/defaultai.cc

```diff
@@ -3330,18 +3330,34 @@ void DefaultAI::check_flag_distances(const Time& gametime) {
 // Dealing with diplomacy actions
 void DefaultAI::diplomacy_actions(const Time& gametime) {
 
+	// don't do any diplomacy for the first 10 + x minutes to avoid click races for allies
+	if (gametime < Time((10 + RNG::static_rand(10)) * 60 * 1000)) {
+		return;
+	}
+
+	const Widelands::PlayerNumber mypn = player_number();
+	const Widelands::Player* me = game().get_player(mypn);
+
+	// if we are defeated or the last one in a team leave team
+	if (me->team_number() != 0 &&
+	    (player_statistics.members_in_team(me->team_number()) == 1 || me->is_defeated())) {
+		game().send_player_diplomacy(mypn, Widelands::DiplomacyAction::kLeaveTeam, 0 /* ignored */);
+		verb_log_dbg_time(gametime, "AI Diplomacy: Player(%d), leaves team (%d) as last one.\n",
+		                  static_cast<unsigned int>(mypn),
+		                  static_cast<unsigned int>(me->team_number()));
+	}
+
+	bool request_accepted = false;
 	for (const Widelands::Game::PendingDiplomacyAction& pda : game().pending_diplomacy_actions()) {
-		if (pda.other == player_number()) {
-			// TODO(Nordfriese): The AI just makes a random choice every time.
-			// In the future, make more strategic decision here. Add asking for alliance
-			bool accept = RNG::static_rand(5) == 0;
-			// don't accept any diplomacy for the first 10 + x minutes to avoid click races for allies
-			accept &= gametime > Time((10 + RNG::static_rand(10)) * 60 * 1000);
-			// only accept if asking player is stronger (based on mil power and land)
-			accept &= player_statistics.get_player_power(pda.sender) *
-			             (player_statistics.get_player_land(pda.sender) / 100) >
-			          player_statistics.get_player_power(player_number()) *
-			             (player_statistics.get_player_land(player_number()) / 100);
+		if (pda.other == mypn) {
+			// accept if diploscore high, else accept only 50%
+			bool accept =
+			   player_statistics.get_diplo_score(pda.sender) >= 30 ||
+			   (player_statistics.get_diplo_score(pda.sender) > 20 && RNG::static_rand(2) == 0);
+			accept = accept &&
+			         player_statistics.members_in_team(pda.sender) <
+			            player_statistics.players_active() - 1 &&
+			         player_statistics.members_in_team(mypn) < player_statistics.players_active() - 1;
```

Comment:
Only the team to be joined needs to be checked. Why prevent leaving the big team? ;)
I hope I didn't mix up the directions.

```suggestion
			accept = accept &&
			         player_statistics.members_in_team(
			            pda.action == Widelands::DiplomacyAction::kInvite ? pda.sender : mypn) <
			         player_statistics.players_active() - 1;
```

Reply:
In this section it is only about accepting, so no leave is prevented. However your solution looks nicer so I'll take it